### PR TITLE
[Bugfix][Rust Frontend] Tolerate NaN-corrupted logprobs in engine-core

### PR DESCRIPTION
**Pull request by almersawi**
## Purpose

Fix a total-outage bug in the Rust frontend's engine-core client: a single request asking for `logprobs`/`prompt_logprobs` while the model produces NaN logits permanently kills the client — `/health` returns 503 and every subsequent request (including ones with no logprobs) fails with `engine-core error: messagepack ext value decode failed: token_ranks must be >= 1 for decoded engine-core logprobs` until process restart.

Rank 0 is an engine output that must be accepted: the sampler computes the rank as `(logprobs >= sampled_logprob).sum(-1)` (`vllm/v1/sample/sampler.py`), which is 0 when the sampled logprob is NaN because every `>=` comparison against NaN is false. The decode failure is frame-scoped (one bad row fails the whole batched `EngineCoreOutputs` payload), happens on the shared output-reader path, and is recorded as a sticky health error, so a per-request data anomaly becomes a server-wide outage.

**Fix:** in `PositionLogprobs::from_decoded_row`, accept and preserve `sampled_rank = 0` along with the engine's original logprob values. This keeps decoding valid for the other requests in the batch. The existing Rust OpenAI serving helper converts NaN and negative infinity to `-9999.0` when constructing HTTP responses; engine-core decoding preserves the numerical results for all consumers.

**Why this does not duplicate existing PRs:**
- #48585 fixes the same *class* of defect (non-finite logprobs reaching the serving layer) but in the **Python** frontend, and for logprob *values*, not ranks. This PR addresses the Rust engine-client rank-zero decode failure; the Rust HTTP helper already handles NaN and negative infinity.
- #50725 is a different code path (process abort from an empty stop string); it only shares the blast-radius pattern.
- #50002 proposed aborting requests on optional NaN detection and was superseded by #50283; this PR handles rank-zero payloads that reach the frontend.
- #45060 establishes all-NaN logits as a real engine state vLLM already handles elsewhere.

Healthy payloads retain their existing behavior. Model evaluation N/A: this changes wire-payload decode tolerance while preserving the engine's numerical output.

AI assistance was used for the original fix and the follow-up changes and tests. The follow-up preserves rank 0 and raw engine values and adds coverage for the existing HTTP conversion.

## Test Plan

Two regression tests cover the layer boundary:

1. `zero_sampled_rank_does_not_fail_frame_mates` decodes a shared `EngineCoreOutputs` batch, preserving rank 0, NaN, and negative infinity in both sample and prompt logprobs while keeping the healthy request intact.
2. `chat_logprobs_clamps_nan_and_negative_infinity_for_json` verifies that the existing OpenAI response conversion emits `-9999.0` for NaN and negative infinity while preserving a finite logprob.

## Test Result

- `cargo nextest run --locked -p vllm-engine-core-client -p vllm-server -E 'package(vllm-engine-core-client) | test(routes::openai::utils::logprobs::tests)'`: **118 passed**.
- `cargo clippy --locked -p vllm-engine-core-client -p vllm-server --all-targets --all-features -- -D warnings`: passed.
- `cargo fmt --all -- --check`, `git diff --check`, and commit-time pre-commit hooks: passed.

Commands ran from `rust/` with a fresh short `TMPDIR` for tests and an existing Cargo target directory. No GPU was required.

Model evaluation N/A: the change preserves the engine's numerical output and adjusts frontend decode tolerance.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. (N/A — internal decode-tolerance fix, no docs affected)
</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
@codex review

**Comment by almersawi** [comment]
@BugenZhao Did you have any chance to check this?

File: rust/src/engine-core-client/src/protocol/logprobs.rs [hunk 1]
```diff
@@ -62,6 +66,14 @@ impl PositionLogprobs {
...
                 index as u32
             };
+            // Map non-finite logprobs (from corrupt logits) to the same -9999.0
+            // sentinel the Python frontend uses, so the serving layer can still
+            // serialize the response.
+            let logprob = if logprob.is_finite() {
+                logprob
+            } else {
+                -9999.0
+            };
```
Comment (BugenZhao) [inline rust/src/engine-core-client/src/protocol/logprobs.rs]: I think this is not necessary? We can preserve the original value within the frontend pipeline, but only clamp it when it's going to be serialized in the serving layer, which we already have: 

https://github.com/vllm-project/vllm/blob/b0898a493737ed34864efb2dbebc36a73ea043c5/rust/src/server/src/routes/openai/utils/logprobs.rs#L294-L296

**Review by BugenZhao — APPROVED** [review]
Rest LGTM. Thanks!
